torch/nn/functional.py

```diff
@@ -5071,6 +5071,10 @@ def multi_head_attention_forward(
         warnings.warn("Byte tensor for key_padding_mask in nn.MultiheadAttention is deprecated. Use bool tensor instead.")
         key_padding_mask = key_padding_mask.to(torch.bool)
 
+    if key_padding_mask is not None and key_padding_mask.dtype != torch.bool:
+        warnings.warn("Non-Boolean tensor for key_padding_mask in nn.MultiheadAttention is not recommended. Use bool tensor instead.")
```

Comment:
To be honest, I'd prefer erroring out on mask types that are not bool or byte, because the authors intent is unclear in this case, and it's likely an error. For the tests that were failing after this change, what's the use case for long mask?